pkg/migration-tool/writer/writer.go

```diff
@@ -0,0 +1,147 @@
+		sigBlockWrite:      sigWrite,
+		migrationJobName:   migrationJobName,
+		progressMetricName: progressMetricName,
+	}
+	return write, nil
+}
+
+// Run runs the remote-writer. It waits for the remote-reader to give access to the in-memory
+// data-block that is written after the most recent fetch. After reading the block and storing
+// the data locally, it gives back the writing access to the remote-reader for further fetches.
+func (rw *RemoteWrite) Run() error {
+	log.Info("msg", "writer is up")
+	for {
+		blockRef, ok := <-rw.sigBlockRead
+		if !ok {
+			break
+		}
+		var buf []byte
+		blockRef.SetDescription(fmt.Sprintf("pushing %.2f...", float64(blockRef.Bytes())/float64(utils.Megabyte)), 1)
+		ps, err := utils.GetorGenerateProgressTimeseries(rw.progressMetricName, rw.migrationJobName)
+		if err != nil {
+			return fmt.Errorf("get or create progress time-series: %w", err)
+		}
+		ps.Append(1, blockRef.Maxt())
+		utils.MergeWithTimeseries(&blockRef.Timeseries, ps.TimeSeries)
+		ts := timeseriesRefToTimeseries(blockRef.Timeseries)
+		if err := rw.sendSamplesWithBackoff(context.Background(), ts, &buf); err != nil {
+			return fmt.Errorf("remote-write run: %w", err)
+		}
+		if err := blockRef.Done(); err != nil {
+			return fmt.Errorf("remote-write: %w", err)
+		}
+		rw.sigBlockWrite <- struct{}{}
+	}
+	log.Info("msg", "writer is down")
+	return nil
+}
+
+// sendSamples to the remote storage with backoff for recoverable errors.
+func (rw *RemoteWrite) sendSamplesWithBackoff(ctx context.Context, samples []prompb.TimeSeries, buf *[]byte) error {
+	// TODO: Add metrics similar to upstream for tool details.
+	req, err := buildWriteRequest(samples, *buf)
+	if err != nil {
+		// Failing to build the write request is non-recoverable, since it will
+		// only error if marshaling the proto to bytes fails.
+		return err
+	}
+
+	backoff := BackOffRetryDuration
+	*buf = req
+
+	attemptStore := func() error {
+		if err := rw.client.Store(ctx, *buf); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		err = attemptStore()
+
+		if err != nil {
+			// If the error is unrecoverable, we should not retry.
+			if _, ok := err.(remote.RecoverableError); !ok {
+				return err
+			}
+
+			// If we make it this far, we've encountered a recoverable error and will retry.
+			time.Sleep(backoff)
+			continue
+		}
+		return nil
+	}
+}
+
+func buildWriteRequest(samples []prompb.TimeSeries, buf []byte) ([]byte, error) {
+	req := &prompb.WriteRequest{
+		Timeseries: samples,
+	}
+	data, err := proto.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// snappy uses len() to see if it needs to allocate a new slice. Make the
+	// buffer as long as possible.
+	if buf != nil {
+		buf = buf[0:cap(buf)]
+	}
+	compressed := snappy.Encode(buf, data)
+	return compressed, nil
+}
+
+func timeseriesRefToTimeseries(tsr []*prompb.TimeSeries) (ts []prompb.TimeSeries) {
```

Comment:
can't say I love this. Why don't we store []prompb.TimeSeries instead of []*prompb.TimeSeries to begin with

Reply:
The response from fetch is promb.QueryResult. Promb.QueryResult has []*Timeseries. This means that by any means, we have to convert. But, converting at the last step (like here) is better, else will slow down the process due to transfering of hugeg data between cpu registers.